pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
-func ctrStop(id string, force bool) error {
-	log.Infof("ctrStop %s %t\n", id, force)
+func ctrStop(containerID string, force bool) error {
+	log.Infof("ctrStop %s %t\n", containerID, force)
 	log.Infof("ctr stop done\n")
 	return nil
 }
 
 // ctrRm remove an existing container
-func ctrRm(id string) error {
-	log.Infof("ctrRm %s\n", id)
+func ctrRm(containerID string) error {
+	log.Infof("ctrRm %s\n", containerID)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return fmt.Errorf("Could not build containerd client. %v", err.Error())
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	container, err := loadContainer(containerID)
+	if err != nil {
+		return err
+	}
+	task, err := container.Task(ctrdCtx, cio.NewAttach(cio.WithStdio))
+	if err != nil {
+		log.Errorf("Could not fetch task of container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Could not fetch task of container: %v. %v", containerID, err.Error())
+	}
+	_, err = task.Delete(ctrdCtx)
+	if err != nil {
+		log.Errorf("Unable to delete task of container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Unable to delete task of container: %v. %v", containerID, err.Error())
+	}
+	err = container.Delete(ctrdCtx)
+	if err != nil {
+		log.Errorf("Unable to delete container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Unable to delete container: %v. %v", containerID, err.Error())
+	}
 	log.Infof("ctrRm done\n")
 	return nil
 }
 
 // ctrCreate create a new container but do not start it
-func ctrCreate(id, imageHash string) error {
-	return nil
+func ctrCreate(containerID, imageName string, status *types.DomainStatus) (string, error) {
+	var (
+		ociOpts       []oci.SpecOpts
+		containerOpts []containerd.NewContainerOpts
+		ociSpec       oci.Spec
+		err           error
+	)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return "", fmt.Errorf("Exception while loading container. %v", err)
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	if len(status.EnvVariables) > 0 {
+		envVars := buildEnvVarsSlice(status.EnvVariables)
+		ociOpts = append(ociOpts, oci.WithEnv(envVars))
+	}
+
+	if len(ociOpts) > 0 {
+		containerOpts = append(containerOpts, containerd.WithSpec(&ociSpec, ociOpts...))
+	}
+	image, err := getImage(imageName)
+	if err != nil {
+		return "", fmt.Errorf("Unable to fetch image %v. %v", imageName, err.Error())
+	}
+
+	containerSnapshot := fmt.Sprintf("%s-snapshot", containerID)
+	containerOpts = append(containerOpts, containerd.WithImage(image))
+	containerOpts = append(containerOpts, containerd.WithNewSnapshot(containerSnapshot, image))
+	containerOpts = append(containerOpts, containerd.WithNewSpec(oci.WithImageConfig(image)))
+	_, err = ctrdClient.NewContainer(
```

Comment:
This is a bit wasteful. We ignore the returned `Container` here, and then recreate it in `ctrCreateNewTask()` below, right down to a new client. Why not just keep the response and use it?

```go
cont, err := ctrdClient.NewContainer(
//...
)
//...
task, err := container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
```

Reply:
Removed.